### PR TITLE
Write spike revising e2e tests

### PR DESCRIPTION
**Pull request by tcharding**
This spike is the result from chatting with everyone on the team trying to work out what is wrong with the e2e tests.

The scope of this spike is simply to try and work out what is wrong with the e2e tests.  Issues will be created (by Tobin) for the next steps.

Fixes: #985

### CODE REVIEW

File: 0011-revise-e2e-test-framework.adoc [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+
+== Suggestions
+
+While the purpose of this spike was to figure out exactly what is wrong with the current end to end tests some valuable suggestions came up, it would be a shame to waste these so adding here for future reference.
+
+* Have the test framework count assertions per test and fail if they don't all run.
+* We may only need to test the happy path for one asset/ledger pair then rely on integration tests for edge cases.
+* Dockerise the tests and run in parallel.
+* Have a tool that shows lines of code touched / not touched during run of the e2e tests (useful for checking that newly added code is exercised).
```
Comment (bonomat) [inline 0011-revise-e2e-test-framework.adoc:65]: I'm not sure if we should try/do this for our e2e tests. This sounds more like a useful unit/integration test feature to me as e2e tests should be blackbox tests and one should write tests that only assert the states, one should not write tests to hit a particular line of code.

**Review by bonomat — COMMENTED** [review]
Thanks for the summary... the hard part is now to come up with something better :D

File: 0011-revise-e2e-test-framework.adoc [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+These are the issues distilled from talking with all team members (excluding Lucas).
+
+=== General
+
+* Tests are not independent.
+* As a developer, I want to be able to run a test against a comit_node/btsieve/docker container that I started myself (e.g. I attached a debugger to cnd), without having to change JS code.
+
+=== Framework
+
+* Not overly happy with mocha but not unhappy either, no real specifics.
```
Comment (D4nte) [inline 0011-revise-e2e-test-framework.adoc:27]: Mocha does not allow parallel run of tests. Jest give you that for free.
We must move to parallel run as the number of tests will explode once we add a 4th asset (which is part of our quarter OKRs).

File: 0011-revise-e2e-test-framework.adoc [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+** from the command line.
+** from within an IDE.
+* Tests should be faster.
+** Each test should be able to be run in parallel.
+
+
+=== Output
+
+Current output for failing tests is lacking because:
+* It is not clear what went wrong.
```
Comment (D4nte) [inline 0011-revise-e2e-test-framework.adoc:51]:
```suggestion

* It is not clear what went wrong.
```
You need a new line to get the list in asciiDoc.

Comment (D4nte) [reply 0011-revise-e2e-test-framework.adoc:51]: Is not? what is not clear?

File: 0011-revise-e2e-test-framework.adoc [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+* There is no information on the state of the two nodes.
+* One needs to know the comit protocol to understand what went wrong.
+* Blockchain errors are hard to understand.
```
Comment (D4nte) [inline 0011-revise-e2e-test-framework.adoc:54]: Do you have some examples? :)

File: 0011-revise-e2e-test-framework.adoc [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+* Not all errors point to the cause i.e., blockchain errors can be caused by an incorrect configuration file.
+
+
+== Suggestions
+
+While the purpose of this spike was to figure out exactly what is wrong with the current end to end tests some valuable suggestions came up, it would be a shame to waste these so adding here for future reference.
+
+* Have the test framework count assertions per test and fail if they don't all run.
```
Comment (D4nte) [inline 0011-revise-e2e-test-framework.adoc:62]: What do you mean?

File: 0011-revise-e2e-test-framework.adoc [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+* We may only need to test the happy path for one asset/ledger pair then rely on integration tests for edge cases.
```
Comment (D4nte) [inline 0011-revise-e2e-test-framework.adoc:63]: I think this requirement goes away if we can parallelise tests

File: 0011-revise-e2e-test-framework.adoc [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+* Dockerise the tests and run in parallel.
+* Have a tool that shows lines of code touched / not touched during run of the e2e tests (useful for checking that newly added code is exercised).
+* Test harness should be called by each test not the other way around (this is probably needed to run tests individually i.e., from an IDE).
+* Consider using testing theory that states individual tests should cleanly: (see Thomas for more information)
+ 1. do setup
+ 2. do stuff
+ 3. make assertions
+* Unit test coverage should be checked as part of a PR, new code requiring before it can be merged.
+* e2e tests should be so clear that outsiders can open them up and see how to code a COMIT bot.
```
Comment (D4nte) [inline 0011-revise-e2e-test-framework.adoc:72]: I think this is an extremely difficult goal especially that we ruled out the usefulness of a client library.

File: 0011-revise-e2e-test-framework.adoc [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+* Assert the casing of JSON keys in all payloads (middleware?) #1138
```
Comment (D4nte) [inline 0011-revise-e2e-test-framework.adoc:73]:
```suggestion
* Assert the casing of JSON keys in all payloads (middleware?) [#1138](https://github.com/comit-network/spikes#1138)
```

**Review by D4nte — COMMENTED** [review]
Thank you for this spike and summarising the issue.
However, I am hungry for solutions now. What's your proposal? How do we move forward from here?

**Comment by D4nte** [comment]
I did put some comments 5 days ago but you requested for review yesterday without addressing/answering them.

File: 0011-revise-e2e-test-framework.adoc (hunk 8 above)
Comment (tcharding) [reply 0011-revise-e2e-test-framework.adoc:73]: Added as suggested.

**Comment by tcharding** [comment]
> I did put some comments 5 days ago but you requested for review yesterday without addressing/answering them.

My bad, I should have messaged this before.  I decided to add a line to the readme pointing the next guy who writes up all the issues to resolve this to read the comments on this issue (that guy is me).

**Comment by tcharding** [comment]
Issue to track implementing the findings of this spike can now be found [here](https://github.com/comit-network/comit-rs/issues/1170)

**Review by D4nte — APPROVED** [review]
